pkg/object/httpserver/mux.go

```diff
@@ -382,52 +386,86 @@ func (m *mux) ServeHTTP(stdw http.ResponseWriter, stdr *http.Request) {
 		return
 	}
 
-	for _, host := range rules.rules {
+	result, path := SearchPath(ctx, rules.rules)
+	switch result {
+	case Found:
+		ci = &cacheItem{ipFilterChan: path.ipFilterChain, path: path}
+		rules.putCacheItem(ctx, ci)
+		m.handleRequestWithCache(rules, ctx, ci)
+	case FoundSkipCache:
+		ci := &cacheItem{ipFilterChan: path.ipFilterChain, path: path}
+		m.handleRequestWithCache(rules, ctx, ci)
+	case MethodNotAllowed:
+		ci := &cacheItem{ipFilterChan: path.ipFilterChain, methodNotAllowed: true}
+		rules.putCacheItem(ctx, ci)
+		m.handleRequestWithCache(rules, ctx, ci)
+	case IPNotAllowed:
+		m.handleIPNotAllow(ctx)
+	case NotFound:
+	default:
+		ci = &cacheItem{ipFilterChan: rules.ipFilterChan, notFound: true}
+		rules.putCacheItem(ctx, ci)
+		m.handleRequestWithCache(rules, ctx, ci)
+	}
+}
+
+const (
+	// NotFound means no path found
+	NotFound SearchResult = "not-found"
+	// IPNotAllowed means context IP is not allowd
+	IPNotAllowed SearchResult = "ip-not-allowed"
+	// MethodNotAllowed means context method is not allowd
+	MethodNotAllowed SearchResult = "method-not-allowed"
+	// Found path
+	Found SearchResult = "found"
+	// FoundSkipCache means found path but skip caching result
+	FoundSkipCache SearchResult = "found-skip-cache"
+)
+
+// SearchPath searches path among list of mux rules
+func SearchPath(ctx context.HTTPContext, rulesToCheck []*muxRule) (SearchResult, *MuxPath) {
+	methodAllowed := false
+	pathFound := false
+	for _, host := range rulesToCheck {
 		if !host.match(ctx) {
 			continue
 		}
 
 		if !host.pass(ctx) {
-			m.handleIPNotAllow(ctx)
-			return
+			return IPNotAllowed, nil
 		}
 
 		for _, path := range host.paths {
 			if !path.matchPath(ctx) {
 				continue
 			}
 
+			// at least one path matches
+			pathFound = true
+
 			if !path.matchMethod(ctx) {
-				ci = &cacheItem{ipFilterChan: path.ipFilterChain, methodNotAllowed: true}
-				rules.putCacheItem(ctx, ci)
-				m.handleRequestWithCache(rules, ctx, ci)
-				return
+				continue
 			}
+			// at least one path has correct method
+			methodAllowed = true
 
 			if !path.pass(ctx) {
-				m.handleIPNotAllow(ctx)
-				return
+				return IPNotAllowed, path
```

Comment:
A little suggestion, our `path matching rule` should be `matchPath + matchMethod + matchHeader`, after the above three conditions are matched, only then `request` is considered to match the `path`, so only when the above conditions are met, should we check whether the `IP is passed`.

```golang
if !path.matchMethod(ctx) {
	continue
}
// at least one path has correct method
methodAllowed = true

var searchResult SearchResult

if path.hasHeaders() {
	if !path.matchHeaders(ctx) {
		continue
	}
	searchResult = FoundSkipCache
} else {
	searchResult = Found
}

if !path.pass(ctx) {
	return IPNotAllowed, nil
}

return searchResult, path
````

Reply:
In this case, we can implement the following rule configuration

```yaml
rules:
  - paths:
    - pathPrefix: /pipeline
      method: POST
      headers:
       X-version: v1
      ip:
        AllowIPs:
         - "8.8.8.8"
      backend: post-v1-demo
    
    - pathPrefix: /pipeline
      method: POST
      headers:
       X-version: v2
      ip:
        AllowIPs:
         - "9.9.9.9"
      backend: post-v2-demo
```

Reply:
Thanks for review @aniaan ! I agree that it's better to have unit tests here so that it would be easier to make these modifications, however currently it would require adding unit test for the whole `httpserver` module.

Reply:
I like your suggestion about routing based on IP and headers.. I think that when the path does not define them, route should match all headers and IPs. If header or IP is defined, it will be used to match and request causes MethodNotAllowed if no correct header or IP is present in the request.

So this will match requests with correct path and method and any IP or any or none header
```yaml
pathPrefix: /pipeline
method: POST
backend: post-v2-demo
```

and in your example it's MethodNotAllowed if request path is correct but header or IP is not present in the request. @aniaan Is this what you mean?

Reply:
Maybe my description is not very accurate, I want to say is that a `path` matching rule is `path + method + headers`, **ip is not part of the path matching**, we have to put the headers check to the front, if the `path + method + headers` are matched to complete, that the user is matching the rule, and then check whether the ip is passed, ip if passed, go backend, not passed, **directly return 403 response**.

<img width="1377" alt="image" src="https://user-images.githubusercontent.com/40813941/157657844-6c018b72-8e69-42de-9444-8f2204be9818.png">